optaplanner-core/src/test/java/org/optaplanner/core/api/solver/SolverManagerTest.java

```diff
@@ -240,4 +246,153 @@ public void skipAhead() throws ExecutionException, InterruptedException {
+        }
+
+        for (SolverJob<TestdataSolution, Long> job : jobs) {
+            assertSolutionInitialized(job.getFinalBestSolution());
+        }
+    }
+
+    /**
+     * Tests whether SolverManager can cope with the situation when more problems than processors are submitted.
+     * In addition, the test verifies whether solving with consumers is affected by this.
+     */
+    @Test(timeout = 600_000)
+    public void moreProblemsThanCpus() throws InterruptedException, ExecutionException {
+        // Use twice the amount of processes than available processors.
+        int problemCount = Runtime.getRuntime().availableProcessors() * 2;
+
+        SolverManager<TestdataSolution, Integer> solverManager = createSolverManagerTestableByDifferentConsumers(problemCount);
+
+        assertSolveWithoutConsumer(problemCount, solverManager);
+        assertSolveWithBestSolutionConsumer(problemCount, solverManager);
+        assertSolveWithFinalBestSolutionConsumer(problemCount, solverManager);
+    }
+
+    private SolverManager<TestdataSolution, Integer> createSolverManagerTestableByDifferentConsumers(int processCount) {
+        final SolverConfig solverConfig = PlannerTestUtils.buildSolverConfig(TestdataSolution.class, TestdataEntity.class)
+                .withPhases(new CustomPhaseConfig().withCustomPhaseCommands(
+                        (ScoreDirector<TestdataSolution> scoreDirector) -> {
+                            TestdataSolution solution = scoreDirector.getWorkingSolution();
+                            TestdataEntity entity = solution.getEntityList().get(0);
+                            scoreDirector.beforeVariableChanged(entity, "value");
+                            entity.setValue(solution.getValueList().get(0));
+                            scoreDirector.afterVariableChanged(entity, "value");
+                            scoreDirector.triggerVariableListeners();
+                        }), new CustomPhaseConfig().withCustomPhaseCommands(
+                        (ScoreDirector<TestdataSolution> scoreDirector) -> {
+                            TestdataSolution solution = scoreDirector.getWorkingSolution();
+                            TestdataEntity entity = solution.getEntityList().get(1);
+                            scoreDirector.beforeVariableChanged(entity, "value");
+                            entity.setValue(solution.getValueList().get(1));
+                            scoreDirector.afterVariableChanged(entity, "value");
+                            scoreDirector.triggerVariableListeners();
+                        }));
+
+        SolverManagerConfig solverManagerConfig = new SolverManagerConfig().withParallelSolverCount(String.valueOf(processCount));
+
+        return SolverManager.create(solverConfig, solverManagerConfig);
+    }
+
+    private void assertSolveWithoutConsumer(int processCount, SolverManager<TestdataSolution, Integer> solverManager) throws InterruptedException, ExecutionException {
```

Comment:
Opposed to the other `assertSolver*` methods, this one lacks any synchronization between submitting the problems to solver manager and checking results. It's a risk in terms of test stability.